test/lib/jdk/test/lib/hotspot/ir_framework/IRMatcher.java

```diff
@@ -0,0 +1,562 @@
+                        append = true; // Append all following lines until we hit the closing </ideal> tag.
+                    }
+                } else if (isPrintOptoAssemblyStart(line)) {
+                    String methodName = getMethodName(compileIdMap, line);
+                    if (methodName != null) {
+                        TestFramework.check(compilations.containsKey(methodName), "Must be second entry of " + methodName);
+                        currentMethod = methodName;
+                        append = true; // Append all following lines until we hit the closing </opto_assembly> tag.
+                    }
+                }
+            }
+        } catch (IOException e) {
+            TestFramework.fail("Error while reading " + hotspotPidFileName, e);
+        }
+    }
+
+    /**
+     * Write the input to the IR method and reset the builder.
+     */
+    private void flushOutput(String line, StringBuilder builder, String currentMethod) {
+        TestFramework.check(!currentMethod.isEmpty(), "current method must be set");
+        IRMethod irMethod = compilations.get(currentMethod);
+        if (line.startsWith("</i")) {
+            // PrintIdeal
+            irMethod.appendIdealOutput(builder.toString());
+        } else {
+            // PrintOptoAssembly
+            irMethod.appendOptoAssemblyOutput(builder.toString());
+        }
+        builder.setLength(0);
+    }
+
+    /**
+     * Only consider non-osr (no "compile_kind") and compilations with C2 (no "level")
+     */
+    private boolean maybeTestEntry(String line) {
+        return line.startsWith("<task_queued") && !line.contains("compile_kind='") && !line.contains("level='");
+    }
+
+    /**
+     * Need to escape XML special characters.
+     */
+    private static void appendLine(StringBuilder builder, String line) {
+        if (line.contains("&")) {
+            line = line.replace("&amp;", "&");
+            line = line.replace("&lt;", "<");
+            line = line.replace("&gt;", ">");
+            line = line.replace("&quot;", "\"");
+            line = line.replace("&apos;", "'");
+        }
+        builder.append(line).append("\n");
+    }
+
+    private static int getCompileId(Matcher matcher) {
+        int compileId = -1;
+        try {
+            compileId = Integer.parseInt(matcher.group(1));
+        } catch (NumberFormatException e) {
+            TestRun.fail("Could not parse compile id", e);
+        }
+        return compileId;
+    }
+
+    /**
+     * Parse the compile id from this line if it belongs to a method that needs to be IR tested (part of test class
+     * and IR encoding from the test VM specifies that this method has @IR rules to be checked).
+     */
+    private void addTestMethodCompileId(Map<Integer, String> compileIdMap, String line) {
+        Pattern pattern = Pattern.compile("compile_id='(\\d+)'.*" + Pattern.quote(testClass.getCanonicalName()) + " (\\S+)");
+        Matcher matcher = pattern.matcher(line);
```

Comment:
similarly to `parseIREncoding`, `pattern` here can be precomputed in `IRMatcher::IRMatcher` and stored into a final instance field.